tests/Functional/ResultCacheTest.php

```diff
@@ -225,6 +225,64 @@ public function testFetchAllSavesCache(): void
         self::assertCount(1, $layerCache->fetch('testcachekey'));
     }
 
+    public function testCacheQueriedOnlyOnceForCacheMiss(): void
+    {
+        $layerCache = $this->createMock(ArrayCache::class);
+        $layerCache->expects(self::once())
+            ->method('fetch')
+            ->willReturn(false);
+
+        $layerCache->expects(self::once())
+            ->method('save');
```

Comment:
Would you recommend to check which values will be saved here?

Reply:
No. It's a functional test, so it should not depend on the internal structure.

We're checking that only one fetch() is executed as it may take significant amount of time to query cache - in case of remote cache server. So from the external point of view, dbal can save anything as long as it fetches it only once per query.

Reply:
> We're checking that only one fetch() is executed

This is the concern of a unit test — cover the implementation details of a given unit. A functional test cannot cover the number of calls to an idempotent method by definition.

Reply:
In the wikipedia definition there is nothing about "cannot cover the number of calls": `Functional testing is a quality assurance process and a type of black-box testing that bases its test cases on the specifications of the software component under test. Functions are tested by feeding them input and examining the output`

The functional test is something that tests the product on the whole (black box), the unit test is intended to test a small part of the product knowing of its internal structure (white box).

`Doctrine\Common\Cache\Cache` is external to `dbal`. It's a complicated input. Another type of such an input is RDBMS. In most functional tests you are counting the number of sql queries that was sent to RDBMS. These sql queries are the output that you are examining.

In our case, we are concerned over the number of cache fetches. So we're counting how many times `dbal` has requested anything from the cache provider. 

Also, `Doctrine\Common\Cache\Cache::fetch()` is not idempotent - even in the simple case of `Memcached` it may be changed by any other process (however, I don't understand why idempotence does matter here).

Reply:
@k0ka please don't mark conversations you did not start as resolved.

Reply:
I can't speak for @morozov but to me, even if the cache component is external to the dbal, it's still an implementation detail and as such, is inside the black box. The user of the black box does not know about it IMO.

Reply:
There already is a discussion about unit and functional attributes of `testCacheQueriedOnlyOnceForCacheMiss` above, which is why I resolve this conversation to prevent talking about similar topics in two different places.